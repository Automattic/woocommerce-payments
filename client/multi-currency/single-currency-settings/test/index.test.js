/**
 * External dependencies
 */
import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

/**
 * Internal dependencies
 */
import SingleCurrencySettings from '../';
import {
	useCurrencies,
	useAvailableCurrencies,
	useDefaultCurrency,
	useEnabledCurrencies,
	useCurrencySettings,
} from 'wcpay/data';

import MultiCurrencySettingsContext from '../../context';

jest.mock( 'wcpay/data', () => ( {
	useCurrencies: jest.fn(),
	useAvailableCurrencies: jest.fn(),
	useDefaultCurrency: jest.fn(),
	useEnabledCurrencies: jest.fn(),
	useCurrencySettings: jest.fn(),
} ) );

const availableCurrencies = {
	USD: {
		code: 'USD',
		rate: 1,
		name: 'United States (US) dollar',
		id: 'usd',
		is_default: true,
		flag: 'ðŸ‡ºðŸ‡¸',
		symbol: '$',
	},
	CAD: {
		code: 'CAD',
		rate: '1.206823',
		name: 'Canadian dollar',
		id: 'cad',
		is_default: false,
		flag: 'ðŸ‡¨ðŸ‡¦',
		symbol: '$',
	},
	GBP: {
		code: 'GBP',
		rate: '0.708099',
		name: 'Pound sterling',
		id: 'gbp',
		is_default: false,
		flag: 'ðŸ‡¬ðŸ‡§',
		symbol: 'Â£',
	},
	EUR: {
		code: 'EUR',
		rate: '0.826381',
		name: 'Euro',
		id: 'eur',
		is_default: false,
		flag: 'ðŸ‡ªðŸ‡º',
		symbol: 'â‚¬',
	},
	AED: {
		code: 'AED',
		rate: '3.6732',
		name: 'United Arab Emirates dirham',
		id: 'aed',
		is_default: false,
		flag: 'ðŸ‡¦ðŸ‡ª',
		symbol: 'Ø¯.Ø¥',
	},
	CDF: {
		code: 'CDF',
		rate: '2000',
		name: 'Congolese franc',
		id: 'cdf',
		is_default: false,
		flag: 'ðŸ‡¨ðŸ‡©',
		symbol: 'Fr',
	},
	NZD: {
		code: 'NZD',
		rate: '1.387163',
		name: 'New Zealand dollar',
		id: 'nzd',
		is_default: false,
		flag: 'ðŸ‡³ðŸ‡¿',
		symbol: '$',
	},
	DKK: {
		code: 'DKK',
		rate: '6.144615',
		name: 'Danish krone',
		id: 'dkk',
		is_default: false,
		flag: 'ðŸ‡©ðŸ‡°',
		symbol: 'DKK',
	},
	BIF: {
		code: 'BIF',
		rate: '1974',
		name: 'Burundian franc',
		id: 'bif',
		is_default: false,
		flag: 'ðŸ‡§ðŸ‡®',
		symbol: 'Fr',
	},
	CLP: {
		code: 'CLP',
		rate: '706.8',
		name: 'Chilean peso',
		id: 'clp',
		is_default: false,
		flag: 'ðŸ‡¨ðŸ‡±',
		symbol: '$',
	},
};

const enabledCurrencies = {
	USD: {
		code: 'USD',
		rate: 1,
		name: 'United States (US) dollar',
		id: 'usd',
		is_default: true,
		flag: 'ðŸ‡ºðŸ‡¸',
		symbol: '$',
	},
	CAD: {
		code: 'CAD',
		rate: '1.42',
		name: 'Canadian dollar',
		id: 'cad',
		is_default: false,
		flag: 'ðŸ‡¨ðŸ‡¦',
		symbol: '$',
	},
	GBP: {
		code: 'GBP',
		rate: '0.708099',
		name: 'Pound sterling',
		id: 'gbp',
		is_default: false,
		flag: 'ðŸ‡¬ðŸ‡§',
		symbol: 'Â£',
	},
	EUR: {
		code: 'EUR',
		rate: '0.826381',
		name: 'Euro',
		id: 'eur',
		is_default: false,
		flag: 'ðŸ‡ªðŸ‡º',
		symbol: 'â‚¬',
	},
};

const defaultCurrency = {
	code: 'USD',
	rate: 1,
	name: 'United States (US) dollar',
	id: 'usd',
	is_default: true,
	flag: 'ðŸ‡ºðŸ‡¸',
	symbol: '$',
};

useCurrencies.mockReturnValue( {
	currencies: {
		available: availableCurrencies,
		enabled: enabledCurrencies,
		default: defaultCurrency,
	},
	isLoading: false,
} );

useAvailableCurrencies.mockReturnValue( availableCurrencies );

useDefaultCurrency.mockReturnValue( defaultCurrency );

useEnabledCurrencies.mockReturnValue( {
	enabledCurrencies: enabledCurrencies,
	submitEnabledCurrenciesUpdate: () => {},
} );

useCurrencySettings.mockReturnValue( {
	currencySettings: {
		EUR: {
			exchange_rate_type: 'automatic',
			manual_rate: null,
			price_charm: null,
			price_rounding: null,
		},
	},
	isLoading: false,
	submitCurrencySettings: jest.fn(),
} );

const containerContext = {
	isSingleCurrencyScreenOpen: true,
	currencyCodeToShowSettingsFor: 'EUR',
	openSingleCurrencySettings: jest.fn(),
	closeSingleCurrencySettings: jest.fn(),
};

const getContainer = () => {
	return render(
		<MultiCurrencySettingsContext.Provider value={ containerContext }>
			<SingleCurrencySettings />
		</MultiCurrencySettingsContext.Provider>
	);
};

describe( 'Single currency settings screen', () => {
	beforeEach( () => {
		jest.clearAllMocks();
		global.wcpaySettings = { zeroDecimalCurrencies: [] };
	} );

	test( 'Page renders correctly', () => {
		const { container } = getContainer();
		expect( container ).toMatchSnapshot();
	} );

	test( 'Settings work correctly', () => {
		getContainer();
		expect(
			screen.queryByText( /Currency Settings/i )
		).toBeInTheDocument();
		expect(
			screen.queryByText( /Current rate: 1 USD = 0.826381 EUR/ )
		).toBeInTheDocument();

		// Check the selects if they have default values.
		expect( screen.getByTestId( 'price_rounding' ) ).toHaveValue( '1' );
		expect( screen.getByTestId( 'price_charm' ) ).toHaveValue( '0' );

		// Test when Automatic rate selected.
		expect(
			screen.getByRole( 'radio', { name: /Fetch rates automatically/ } )
		).toBeChecked();
		expect( screen.queryByTestId( 'manual_rate_input' ) ).toBeNull();
		expect( screen.getByTestId( 'store_currency_value' ) ).toBeVisible();
		fireEvent.change( screen.getByTestId( 'store_currency_value' ), {
			target: {
				value: '10',
			},
		} );
		expect( screen.getByTestId( 'calculated_value' ) ).toHaveTextContent(
			'8.00â‚¬'
		);
		// Manual Rate calculation.
		fireEvent.click( screen.getByRole( 'radio', { name: /Manual/ } ) );
		expect( screen.getByTestId( 'manual_rate_input' ) ).toBeVisible();
		fireEvent.change( screen.getByTestId( 'manual_rate_input' ), {
			target: {
				value: '2',
			},
		} );

		expect( screen.getByTestId( 'calculated_value' ) ).toHaveTextContent(
			'20.00â‚¬'
		);

		fireEvent.change( screen.getByTestId( 'manual_rate_input' ), {
			target: {
				value: '1.77',
			},
		} );

		expect( screen.getByTestId( 'calculated_value' ) ).toHaveTextContent(
			'17.00â‚¬'
		);

		// Price rounding calculation.
		fireEvent.change( screen.getByTestId( 'price_rounding' ), {
			target: { value: '0.5' },
		} );
		expect( screen.getByTestId( 'calculated_value' ) ).toHaveTextContent(
			'17.50â‚¬'
		);

		// Price charm calculation.
		fireEvent.change( screen.getByTestId( 'price_charm' ), {
			target: { value: '-0.05' },
		} );
		expect( screen.getByTestId( 'calculated_value' ) ).toHaveTextContent(
			'17.45â‚¬'
		);

		// Submit settings test.
		const { submitCurrencySettings } = useCurrencySettings();

		fireEvent.click(
			screen.getByRole( 'button', { name: /Save Changes/i } )
		);

		expect( submitCurrencySettings ).toHaveBeenCalledWith( 'EUR', {
			exchange_rate_type: 'manual',
			manual_rate: '1.77',
			price_rounding: '0.5',
			price_charm: '-0.05',
		} );
	} );
} );
