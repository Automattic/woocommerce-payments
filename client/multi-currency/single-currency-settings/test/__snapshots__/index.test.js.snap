// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single currency settings screen Page renders correctly 1`] = `
<div>
  <div
    class="single-currency-settings"
  >
    <div
      class="wcpay-settings-layout"
    >
      <h2
        class="single-currency-settings-breadcrumb"
      >
        <button
          class="components-button is-link"
          type="button"
        >
          Enabled currencies
        </button>
         
        &gt; 
        Euro
         (
        EUR
        )
         
        ðŸ‡ªðŸ‡º
      </h2>
      <div
        class="settings-section"
      >
        <div
          class="settings-section__details"
        >
          <h2>
            Currency settings
          </h2>
          <p>
            Choose between automatic or manual exchange rates and modify formatting rules to refine the display of your prices.
          </p>
        </div>
        <div
          class="settings-section__controls"
        >
          <div
            class="components-surface components-card single-currency-settings-currency-settings css-nsno0f-View-Surface-getBorders-primary-Card-rounded em57xhy0"
            data-wp-c16t="true"
            data-wp-component="Card"
          >
            <div
              class="css-mgwsf4-View-Content em57xhy0"
            >
              <div
                class="components-card__body components-card-body css-1nwhnu3-View-Body-borderRadius-medium em57xhy0"
                data-wp-c16t="true"
                data-wp-component="CardBody"
              >
                <h3>
                  Exchange rate
                </h3>
                <fieldset>
                  <ul>
                    <li>
                      <label
                        for="single-currency-settings__automatic_rate_radio"
                      >
                        <div
                          class="single-currency-settings-radio-wrapper"
                        >
                          <input
                            checked=""
                            class="exchange-rate-selector"
                            id="single-currency-settings__automatic_rate_radio"
                            name="wcpay_multi_currency_exchange_rate_aud"
                            type="radio"
                            value="automatic"
                          />
                          <h4>
                            Fetch rates automatically
                          </h4>
                        </div>
                        <p
                          class="single-currency-settings-description single-currency-settings-description-inset"
                        >
                          Current rate: 1 USD = 0.826381 EUR (Last updated: September 24, 2021 01:14)
                        </p>
                      </label>
                    </li>
                    <li>
                      <label
                        for="single-currency-settings__manual_rate_radio"
                      >
                        <div
                          class="single-currency-settings-radio-wrapper"
                        >
                          <input
                            class="exchange-rate-selector"
                            id="single-currency-settings__manual_rate_radio"
                            name="wcpay_multi_currency_exchange_rate_aud"
                            type="radio"
                            value="manual"
                          />
                          <h4>
                            Manual
                          </h4>
                        </div>
                        <p
                          class="single-currency-settings-description single-currency-settings-description-inset"
                        >
                          Enter your fixed rate of exchange
                        </p>
                      </label>
                    </li>
                    
                  </ul>
                </fieldset>
                <h3>
                  Formatting rules
                </h3>
                <fieldset>
                  <ul>
                    <li>
                      <h4>
                        Price rounding
                      </h4>
                      <select
                        data-testid="price_rounding"
                      >
                        <option
                          value="0"
                        >
                          None
                        </option>
                        <option
                          value="0.25"
                        >
                          0.25
                        </option>
                        <option
                          value="0.5"
                        >
                          0.50
                        </option>
                        <option
                          value="1"
                        >
                          1.00 (recommended)
                        </option>
                        <option
                          value="5"
                        >
                          5.00
                        </option>
                        <option
                          value="10"
                        >
                          10.00
                        </option>
                      </select>
                      <p
                        class="single-currency-settings-description"
                      >
                        Make your EUR prices consistent by rounding them up after they're converted.
                        <button
                          class="components-button is-link"
                          type="button"
                        >
                          Learn more
                        </button>
                      </p>
                    </li>
                    <li>
                      <h4>
                        Charm pricing
                      </h4>
                      <select
                        data-testid="price_charm"
                      >
                        <option
                          value="0"
                        >
                          None
                        </option>
                        <option
                          value="-0.01"
                        >
                          -0.01
                        </option>
                        <option
                          value="-0.05"
                        >
                          -0.05
                        </option>
                      </select>
                      <p
                        class="single-currency-settings-description"
                      >
                        Reduce the converted price for a specific amount
                        <button
                          class="components-button is-link"
                          type="button"
                        >
                          Learn more
                        </button>
                      </p>
                    </li>
                  </ul>
                </fieldset>
              </div>
            </div>
            <div
              aria-hidden="true"
              class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
              data-wp-c16t="true"
              data-wp-component="Elevation"
            />
            <div
              aria-hidden="true"
              class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
              data-wp-c16t="true"
              data-wp-component="Elevation"
            />
          </div>
        </div>
      </div>
      <div
        class="settings-section"
      >
        <div
          class="settings-section__details"
        >
          <h2>
            Preview
          </h2>
          <p>
            Enter a price in your default currency (United States (US) dollar) to see it converted to Euro using the exchange rate and formatting rules above.
          </p>
        </div>
        <div
          class="settings-section__controls"
        >
          <div
            class="single-currency-settings-preview-wrapper"
          >
            <div
              class="components-surface components-card css-nsno0f-View-Surface-getBorders-primary-Card-rounded em57xhy0"
              data-wp-c16t="true"
              data-wp-component="Card"
            >
              <div
                class="css-mgwsf4-View-Content em57xhy0"
              >
                <div
                  class="components-card__body components-card-body css-1nwhnu3-View-Body-borderRadius-medium em57xhy0"
                  data-wp-c16t="true"
                  data-wp-component="CardBody"
                >
                  <div>
                    <h4>
                      United States (US) dollar
                    </h4>
                    <div>
                      <div
                        class="components-base-control with-value css-wdf2ti-Wrapper ej5x27r4"
                      >
                        <div
                          class="components-base-control__field css-10urnh1-StyledField-deprecatedMarginField ej5x27r3"
                        >
                          <div
                            class="text-control-with-affixes text-control-with-suffix"
                          >
                            <input
                              aria-describedby="inspector-text-control-with-affixes-0__suffix"
                              class="components-text-control__input"
                              data-testid="store_currency_value"
                              id="inspector-text-control-with-affixes-0"
                              type="text"
                              value="20"
                            />
                            <span
                              class="text-control-with-affixes__suffix"
                              id="inspector-text-control-with-affixes-0__suffix"
                            >
                              $
                            </span>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div>
                    <h4>
                      Euro
                    </h4>
                    <div>
                      <strong
                        data-testid="calculated_value"
                      >
                        â‚¬17.00
                      </strong>
                    </div>
                  </div>
                </div>
              </div>
              <div
                aria-hidden="true"
                class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
                data-wp-c16t="true"
                data-wp-component="Elevation"
              />
              <div
                aria-hidden="true"
                class="components-elevation css-91yjwm-View-Elevation-sx-Base-elevationClassName em57xhy0"
                data-wp-c16t="true"
                data-wp-component="Elevation"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="settings-section single-currency-settings-save-settings-section"
      >
        <div
          class="settings-section__details"
        />
        <div
          class="settings-section__controls"
        >
          <button
            class="components-button is-primary"
            type="button"
          >
            Save changes
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
